// vite-plugin-stylelint/dist/utils-C9YzWEY8.mjs
import { createFilter, normalizePath } from "@rollup/pluginutils";
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJSMin = (cb, mod) => () => (mod || cb((mod = { exports: {} }).exports, mod), mod.exports);
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
//#region src/constants.ts
const STYLELINT_SEVERITY = {
	ERROR: "error",
	WARNING: "warning"
};
const PLUGIN_NAME = "vite:stylelint";
const COLOR_MAPPING = {
	error: "red",
	warning: "yellow",
	plugin: "magenta"
};
//#endregion
//#region src/utils.ts
var import_picocolors = /* @__PURE__ */ __toESM((/* @__PURE__ */ __commonJSMin(((exports, module) => {
	let p = process || {}, argv = p.argv || [], env = p.env || {};
	let isColorSupported = !(!!env.NO_COLOR || argv.includes("--no-color")) && (!!env.FORCE_COLOR || argv.includes("--color") || p.platform === "win32" || (p.stdout || {}).isTTY && env.TERM !== "dumb" || !!env.CI);
	let formatter = (open, close, replace = open) => (input) => {
		let string = "" + input, index = string.indexOf(close, open.length);
		return ~index ? open + replaceClose(string, close, replace, index) + close : open + string + close;
	};
	let replaceClose = (string, close, replace, index) => {
		let result = "", cursor = 0;
		do {
			result += string.substring(cursor, index) + replace;
			cursor = index + close.length;
			index = string.indexOf(close, cursor);
		} while (~index);
		return result + string.substring(cursor);
	};
	let createColors = (enabled = isColorSupported) => {
		let f = enabled ? formatter : () => String;
		return {
			isColorSupported: enabled,
			reset: f("\x1B[0m", "\x1B[0m"),
			bold: f("\x1B[1m", "\x1B[22m", "\x1B[22m\x1B[1m"),
			dim: f("\x1B[2m", "\x1B[22m", "\x1B[22m\x1B[2m"),
			italic: f("\x1B[3m", "\x1B[23m"),
			underline: f("\x1B[4m", "\x1B[24m"),
			inverse: f("\x1B[7m", "\x1B[27m"),
			hidden: f("\x1B[8m", "\x1B[28m"),
			strikethrough: f("\x1B[9m", "\x1B[29m"),
			black: f("\x1B[30m", "\x1B[39m"),
			red: f("\x1B[31m", "\x1B[39m"),
			green: f("\x1B[32m", "\x1B[39m"),
			yellow: f("\x1B[33m", "\x1B[39m"),
			blue: f("\x1B[34m", "\x1B[39m"),
			magenta: f("\x1B[35m", "\x1B[39m"),
			cyan: f("\x1B[36m", "\x1B[39m"),
			white: f("\x1B[37m", "\x1B[39m"),
			gray: f("\x1B[90m", "\x1B[39m"),
			bgBlack: f("\x1B[40m", "\x1B[49m"),
			bgRed: f("\x1B[41m", "\x1B[49m"),
			bgGreen: f("\x1B[42m", "\x1B[49m"),
			bgYellow: f("\x1B[43m", "\x1B[49m"),
			bgBlue: f("\x1B[44m", "\x1B[49m"),
			bgMagenta: f("\x1B[45m", "\x1B[49m"),
			bgCyan: f("\x1B[46m", "\x1B[49m"),
			bgWhite: f("\x1B[47m", "\x1B[49m"),
			blackBright: f("\x1B[90m", "\x1B[39m"),
			redBright: f("\x1B[91m", "\x1B[39m"),
			greenBright: f("\x1B[92m", "\x1B[39m"),
			yellowBright: f("\x1B[93m", "\x1B[39m"),
			blueBright: f("\x1B[94m", "\x1B[39m"),
			magentaBright: f("\x1B[95m", "\x1B[39m"),
			cyanBright: f("\x1B[96m", "\x1B[39m"),
			whiteBright: f("\x1B[97m", "\x1B[39m"),
			bgBlackBright: f("\x1B[100m", "\x1B[49m"),
			bgRedBright: f("\x1B[101m", "\x1B[49m"),
			bgGreenBright: f("\x1B[102m", "\x1B[49m"),
			bgYellowBright: f("\x1B[103m", "\x1B[49m"),
			bgBlueBright: f("\x1B[104m", "\x1B[49m"),
			bgMagentaBright: f("\x1B[105m", "\x1B[49m"),
			bgCyanBright: f("\x1B[106m", "\x1B[49m"),
			bgWhiteBright: f("\x1B[107m", "\x1B[49m")
		};
	};
	module.exports = createColors();
	module.exports.createColors = createColors;
})))(), 1);
const getOptions = ({ test, dev, build, cache, include, exclude, stylelintPath, formatter, lintInWorker, lintOnStart, lintDirtyOnly, emitError, emitErrorAsWarning, emitWarning, emitWarningAsError, ...stylelintOptions }) => ({
	test: test ?? false,
	dev: dev ?? true,
	build: build ?? false,
	cache: cache ?? true,
	include: include ?? ["src/**/*.{css,scss,sass,less,styl,vue,svelte}"],
	exclude: exclude ?? ["node_modules", "virtual:"],
	stylelintPath: stylelintPath ?? "stylelint",
	formatter: formatter ?? "string",
	lintInWorker: lintInWorker ?? false,
	lintOnStart: lintOnStart ?? false,
	lintDirtyOnly: lintDirtyOnly ?? true,
	emitError: emitError ?? true,
	emitErrorAsWarning: emitErrorAsWarning ?? false,
	emitWarning: emitWarning ?? true,
	emitWarningAsError: emitWarningAsError ?? false,
	...stylelintOptions
});
const getFilter = (options) => createFilter(options.include, options.exclude);
const initializeStylelint = async (options) => {
	const { stylelintPath, formatter } = options;
	try {
		const module = await import(stylelintPath);
		const stylelintInstance = module?.default ?? module;
		return {
			stylelintInstance,
			formatter: typeof formatter === "string" ? await stylelintInstance.formatters[formatter] : formatter
		};
	} catch (error) {
		throw new Error(`Failed to initialize Stylelint. Have you installed and configured correctly? ${error}`);
	}
};
const getStylelintLinterOptions = (options) => ({
	...options,
	allowEmptyInput: true
});
const isVirtualModule = (id) => id.startsWith("virtual:") || id[0] === "\0" || !id.includes("/");
const getFilePath = (id) => normalizePath(id).split("?")[0];
const shouldIgnoreModule = (id, filter) => {
	if (isVirtualModule(id)) return true;
	if (!filter(id)) return true;
	const filePath = getFilePath(id);
	if ([".vue", ".svelte"].some((extname) => filePath.endsWith(extname))) return !(id.includes("?") && id.includes("type=style"));
	return false;
};
const colorize = (text, textType) => import_picocolors.default[COLOR_MAPPING[textType]](text);
const log = (text, textType, context) => {
	console.log("");
	if (context) {
		if (textType === "error") context.error(text);
		else if (textType === "warning") context.warn(text);
	} else console.log(`${text}  Plugin: ${colorize(PLUGIN_NAME, "plugin")}\r\n`);
};
const lintFiles = async ({ files, stylelintInstance, formatter, options }, context) => await stylelintInstance.lint({
	...getStylelintLinterOptions(options),
	files
}).then((linterResult) => {
	if (!linterResult || linterResult.results.length === 0) return;
	let results = linterResult.results.filter((item) => !item.ignored);
	if (!options.emitError) {
		results = results.map((item) => ({
			...item,
			warnings: item.warnings.filter((warning) => warning.severity !== STYLELINT_SEVERITY.ERROR)
		}));
		linterResult.errored = false;
	}
	if (!options.emitWarning) results = results.map((item) => ({
		...item,
		warnings: item.warnings.filter((warning) => warning.severity !== STYLELINT_SEVERITY.WARNING)
	}));
	results = results.filter((item) => item.warnings.length > 0);
	if (results.length === 0) return;
	linterResult.results = results;
	return log(formatter(results, linterResult), (() => {
		if (linterResult.errored) return options.emitErrorAsWarning ? "warning" : "error";
		return options.emitWarningAsError ? "error" : "warning";
	})(), context);
});
//#endregion
export { lintFiles as a, initializeStylelint as i, getFilter as n, shouldIgnoreModule as o, getOptions as r, PLUGIN_NAME as s, getFilePath as t };
